_examples/hellocolibri/hellocolibri.go

```diff
@@ -0,0 +1,340 @@
+	//
+	// find the server address
+	pathquerier := sciond.Querier{
+		Connector: daemon,
+		IA:        localIA,
+	}
+	pathsToDst, err := pathquerier.Query(ctx, dstIA)
+	check(err)
+	p := pathsToDst[0]
+	fmt.Printf("Will connect to server using path %s\n", p)
+
+	serverUdpAddr, err := net.ResolveUDPAddr("udp", serverSciondPath)
+	check(err)
+	serverUdpAddr.Port = serverPort
+	serverAddr := &snet.UDPAddr{
+		IA:      dstIA,
+		Path:    p.Path(),
+		Host:    serverUdpAddr,
+		NextHop: p.UnderlayNextHop(),
+	}
+	// connect to the server
+	udpAddr, err := net.ResolveUDPAddr("udp", clientSciondPath)
+	check(err)
+	udpAddr.Port = 44321 // TODO(juagargi) or zero?
+
+	// connect to server
+	conn, err := scionNet.Dial(ctx, "udp", udpAddr, serverAddr, addr.SvcNone)
+	check(err)
+	_, err = conn.Write([]byte("hello there, best effort"))
+	check(err)
+	data := <-serverChan
+	fmt.Printf("we know the server got data: %s\n", string(data))
+	conn.Close()
+	//
+	//
+	// now dial using colibri
+	cp := &colpath.ColibriPath{}
+	err = cp.DecodeFromBytes(colibriPath.Raw)
+	check(err)
+	fmt.Printf("We will send traffic using colibri now, with path %s\n", printPath(cp))
+	serverAddr.Path.Type = colpath.PathType
+	serverAddr.Path.Raw = colibriPath.Raw
+
+	conn, err = scionNet.Dial(ctx, "udp", udpAddr, serverAddr, addr.SvcNone)
+	check(err)
+	_, err = conn.Write([]byte("hello there, colibri carried data"))
+	check(err)
+	data = <-serverChan
+	fmt.Printf("we know the server got data: %s\n", string(data))
+	conn.Close()
+
+	// clean the last reservation obtained
+	err = daemon.ColibriCleanupRsv(ctx, &setupReq2.Id, setupReq2.Index)
+	check(err)
+	fmt.Println("Reservation cleaned up")
+}
+
+func clientExtendedAPI(serverChan chan []byte) {
+	fmt.Println("client started")
+	ctx, cancelF := context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancelF()
+
+	localUdpAddr, err := net.ResolveUDPAddr("udp", clientSciondPath)
+	check(err)
+	localUdpAddr.Port = 44321 // TODO(juagargi) or zero?
+	dstUdpAddr, err := net.ResolveUDPAddr("udp", serverSciondPath)
+	check(err)
+	dstUdpAddr.Port = serverPort
+	dstIA, err := addr.IAFromString("1-ff00:0:112")
+	check(err)
+
+	daemon, err := sciond.NewService(clientSciondPath).Connect(ctx)
+	check(err)
+	localIA, err := daemon.LocalIA(ctx)
+	check(err)
+	dispatcher := reliable.NewDispatcher(reliable.DefaultDispPath)
+	scionNet := snet.NewNetwork(localIA, dispatcher, sciond.RevHandler{Connector: daemon})
+	dstAddr := &snet.UDPAddr{
+		IA:   dstIA,
+		Host: dstUdpAddr,
+	}
+
+	fmt.Println("deleteme 1")
+	rsv, err := colapi.NewReservation(ctx, scionNet, daemon, dstAddr, 9, 0, func(a, b libcol.FullTrip) bool {
+		// sorting function, return true if a < b. The first trip will be the chosen one
+		return a.ExpirationTime().Before(b.ExpirationTime())
+	})
+	check(err)
+	fmt.Println("deleteme 2")
+
+	// auto renew reservation
+	err = rsv.Open(ctx, localUdpAddr, func(r *colapi.Reservation, err error) {
```

Comment:
The way the API is design forces the user to not make mistakes: since every renewal changes the path, the path itself is contained and managed inside the object, private to the user.
I can add a call to retrieve a copy of the path, but how would you use it? You mention you'd like to upgrade an existing `snet.Conn`. Can you give me a use case for this? (maybe I have overseen something in my scenarios). There is (will be) a special case for QUIC, in where I must be able to construct a quic socket with a reservation, but this is left out for later.

Regarding `Listen`: I don't even think it makes sense to `Read` from a reservation, but I kept the call; I should probably remove it. A call to `Listen` would receive the (not yet implemented) call from `sciond` to authorize the reservation or not. But once the reservation is set up, it will simply receive packets on that socket (IP-port), and _any regular_ call to `Listen` will work (e.g. the server in `hellocolibri`). Do you think they should fail to accept packets from a reservation?

Reply:
> The way the API is design forces the user to not make mistakes: since every renewal changes the path, the path itself is contained and managed inside the object, private to the user.

That exact behavior could be encapsulated inside a `snet.Path` though as I see it? The "actual" path (the bytes to use in packet headers) is returned from the `Path()` getter on `snet.Path` and can hence be updated whenever the reservation is renewed.

> I can add a call to retrieve a copy of the path, but how would you use it? You mention you'd like to upgrade an existing `snet.Conn`. Can you give me a use case for this? (maybe I have overseen something in my scenarios). There is (will be) a special case for QUIC, in where I must be able to construct a quic socket with a reservation, but this is left out for later.

As for use-cases: Generally Colibri only makes much sense for connection-oriented traffic, right? So stuff like QUIC and RTP/WebRTC. Here are two use cases using either protocol:
- A file download using QUIC: A client dials into a server and initiates a download. In this case the client is not interested in having guaranteed bandwidth to the server, as all the client needs to do is send the initial request (e.g. HTTP GET) and send QUIC ACKs. The server sending the file to the client, however, might want to send using Colibri. In this case the server (=listener) accepts a new client connection and would then have to upgrade the connection to the client to Colibri after the connection was already established.
- A WebRTC call as in my thesis. Both peers in a call are basically equal and there is no big client-server distinction one a call is running. However, the underlying call setup may involve one peer using a listener and the other peer dialing into that listener, or using STUN to patch together two dialed connections. In either case, it should be possible for either a dialed connection or a bound listening connection to upgrade to Colibri. As in the previous example, a bound listening connection can only be upgraded to Colibri after the connection is already established. In case of using STUN, a dialed connection also needs to have the ability to switch to Colibri after initially not using Colibri, since STUN involves first connecting to a STUN server, then re-using the same connection to continue talking to the call peer (so it actually involves a change of the destination).

I would imagine using path processors to choose Colibri paths:
- There might be a case where a switch away from Colibri back to regular SCION is wanted. E.g. the only available Colibri path has huge latency because it takes a detour, and causes bad QoS/QoE. Switching to a more direct SCION path would provide lower latency.
- Or simply a switch from one Colibri path to another (e.g. initial choice was bad), all without having to close and create new connections.

Also creating a reservation involves another request which would increase the time to connectivity. Therefore I think it would make a lot of sense to start many Colibri connections as regular SCION connections and parallelize the initial packet sending over SCION with the requesting of an e2e reservation. Once the e2e reservation is done and a Colibri path is available, the connection could be upgraded to use the Colibri path.

> Regarding Listen: I don't even think it makes sense to Read from a reservation, but I kept the call; I should probably remove it.

Why not `Read()`? `Open` simply dials a connection and it needs to be possible to read and write packets just like on a regular scion connection. Or what do you mean?

> A call to `Listen` would receive the (not yet implemented) call from `sciond` to authorize the reservation or not. But once the reservation is set up, it will simply receive packets on that socket (IP-port), and _any regular_ call to `Listen` will work (e.g. the server in `hellocolibri`). Do you think they should fail to accept packets from a reservation?

To make sure I understand correctly: If a listener accepts an e2e reservation, then is closed, and another listener is re-opened on the same port, I guess inherently by the design of Colibri the admission provided by the previous listener on the same port stays valid? Only renewals would then be handled by the new admission callback used by the new listener.

It might make sense to go for the more cautious approach. Who knows what kind of shenanigans could otherwise become possible. Each listener can keep a record over admitted Colibri clients – if it receives an incoming packet that uses a Colibri path and the client is not listed as having been admitted via the listener's admission callback, the packet could be dropped/rejected. This would happen when a now closed listener on the same port previously accepted the reservation. The client would have to first request a new e2e reservation that needs to be accepted by the new listener on the same port.

Reply:
> The way the API is design forces the user to not make mistakes: since every renewal changes the path, the path itself is contained and managed inside the object, private to the user.
I can add a call to retrieve a copy of the path, but how would you use it? You mention you'd like to upgrade an existing snet.Conn. Can you give me a use case for this? (maybe I have overseen something in my scenarios). There is (will be) a special case for QUIC, in where I must be able to construct a quic socket with a reservation, but this is left out for later.

The same problem exists for SCION paths that also expire (not quite so quickly, but the issue is identical). This is not addressed in snet etc. but is left to the applications. Imho, a high level API as proposed here should not be specific to colibri, but should be more generic and integrate the logic for renewal, fallbacks etc in one package. I agree with Jonas that enabling/disabling the use of Colibri for an existing connection seems like an important feature, not least because of the potential need to fallback from Colibri to best-effort SCION.

My suggestion would be to add colibri support to "pan"  (see #187, which is still incomplete and has been collecting dust, unfortunately) instead of building a separate high-level API for Colibri only.

Reply:
> That exact behavior could be encapsulated inside a `snet.Path` though as I see it? The "actual" path (the bytes to use in packet headers) is returned from the `Path()` getter on `snet.Path` and can hence be updated whenever the reservation is renewed.

I just made the `Reservation` implement `snet.Path`: it can now be used directly if so wanted.

> As for use-cases: Generally Colibri only makes much sense for connection-oriented traffic, right?

No no, it makes sense also for connectionless protocols.

> So stuff like QUIC and RTP/WebRTC. Here are two use cases using either protocol:
> 
> * A file download using QUIC: A client dials into a server and initiates a download. In this case the client is not interested in having guaranteed bandwidth to the server, as all the client needs to do is send the initial request (e.g. HTTP GET) and send QUIC ACKs. The server sending the file to the client, however, might want to send using Colibri. In this case the server (=listener) accepts a new client connection and would then have to upgrade the connection to the client to Colibri after the connection was already established.

Keep in mind that in this scenario, the server will setup a new reservation, meaning that the client must be able to accept it, either via whitelists/blacklists or via a callback to grant the setup. The same applies for renewals.
But I understand the scenario. With the current API, the server gets the destination from the incoming connection, creates a reservation, and opens it. It then proceeds to write packets to the `Reservation`.

> * A WebRTC call as in my thesis. Both peers in a call are basically equal and there is no big client-server distinction one a call is running. However, the underlying call setup may involve one peer using a listener and the other peer dialing into that listener, or using STUN to patch together two dialed connections. In either case, it should be possible for either a dialed connection or a bound listening connection to upgrade to Colibri. As in the previous example, a bound listening connection can only be upgraded to Colibri after the connection is already established. In case of using STUN, a dialed connection also needs to have the ability to switch to Colibri after initially not using Colibri, since STUN involves first connecting to a STUN server, then re-using the same connection to continue talking to the call peer (so it actually involves a change of the destination).

I am unfamiliar with STUN, but I believe that this scenario is accepted by the current API: client A dials to a STUN server, which in turn dials to a peer B. After that, client A opens a reservation to the destination, which must be the STUN server, and starts using that `Reservation`.

In general, the action flow for upgrading a connection to colibri would look like this:
- client has an existing, non colibri, connection to another peer that we will call server
- client sends data to the server using the non colibri connection
- client creates a new reservation to server, still can send data to server via the non colibri connection
- client opens the colibri reservation
- client can now send data using the colibri `Reservation` or the non colibri connection, or both. 

Only caveat here is the `dispatcher`: we cannot register to listening sockets using the same port, as they would clash.
This also makes even more outstanding the fact that receiving packets on a colibri connection is no different at all compared to regular scion packets, when looking from the perspective of the listener.

> I would imagine using path processors to choose Colibri paths:
> 
> * There might be a case where a switch away from Colibri back to regular SCION is wanted. E.g. the only available Colibri path has huge latency because it takes a detour, and causes bad QoS/QoE. Switching to a more direct SCION path would provide lower latency.

Yes, this would be desirable. Again, the abstraction let us not stop here, as hidden paths or epic are also paths that could be selected from a path processor. All of them could be seen as `snet.Path`

> * Or simply a switch from one Colibri path to another (e.g. initial choice was bad), all without having to close and create new connections.

With the current API + `Reservation` been an `snet.Path` you don't need to close connections.
But once again, a caveat: if closing a connection or establishing it imply changes to the internal state of the connection, we might have to do it anyways. E.g. QUIC should have no flow control other than constant bit rate when the underlying connection is colibri.
When closing and opening a connection is no operation, the result of switching back and forth between colibri and regular scion is accepted by the current API, even without the `snet.Path` interface.

> Also creating a reservation involves another request which would increase the time to connectivity. Therefore I think it would make a lot of sense to start many Colibri connections as regular SCION connections and parallelize the initial packet sending over SCION with the requesting of an e2e reservation. Once the e2e reservation is done and a Colibri path is available, the connection could be upgraded to use the Colibri path.
> 
> Why not `Read()`? `Open` simply dials a connection and it needs to be possible to read and write packets just like on a regular scion connection. Or what do you mean?

`Read` gives the impression that this colibri connection can be used bidirectionally, which is not true. In can be used to read packets, but with a regular call to `Listen` and `Read` it would be the same. The call is in the API because removing it would prevent easily reading back e.g. ACKs and short responses, so I guess it will have to be very well documented to remove the sensation of reservation when reading. Note that for the same reasons, there is no call to `Listen`.

> To make sure I understand correctly: If a listener accepts an e2e reservation, then is closed, and another listener is re-opened on the same port, I guess inherently by the design of Colibri the admission provided by the previous listener on the same port stays valid? Only renewals would then be handled by the new admission callback used by the new listener.

The mechanism of how to accept or deny a setup/renewal by the endhost is not decided yet. Originally it was done via white/black lists, but I would love to have a design in which we have a callback.
That been said, the listener allowing/denying connection setups would always be present, and a new "ingester" would be spawned (maybe on a different port) if the setup was accepted. E.g. via a response back from the "accepter" with a new port, or a different mechanism.

> It might make sense to go for the more cautious approach. Who knows what kind of shenanigans could otherwise become possible. Each listener can keep a record over admitted Colibri clients – if it receives an incoming packet that uses a Colibri path and the client is not listed as having been admitted via the listener's admission callback, the packet could be dropped/rejected. This would happen when a now closed listener on the same port previously accepted the reservation. The client would have to first request a new e2e reservation that needs to be accepted by the new listener on the same port.

The mechanism in colibri forbids a reservation from being created or renewed if the endhost doesn't accept it. So the listener would never see colibri packets if it didn't accept the reservation. 
Please remember that this exact part is still a TODO in the code, as we are discussing here how we want to communicate the endhost with the service.

Reply:
> The same problem exists for SCION paths that also expire (not quite so quickly, but the issue is identical). This is not addressed in snet etc. but is left to the applications. Imho, a high level API as proposed here should not be specific to colibri, but should be more generic and integrate the logic for renewal, fallbacks etc in one package. I agree with Jonas that enabling/disabling the use of Colibri for an existing connection seems like an important feature, not least because of the potential need to fallback from Colibri to best-effort SCION.

Partially agree: unifying all paths will also expose differences: the high level actions that make sense for colibri might not make sense for others. E.g. a fallback in colibri doesn't make sense unless it can be described to the service itself. It then must be characterized somehow by a description such as minimum reservation bandwidth X iff same latency, but X-2 otherwise, etc. Or send a list of reservations to the service, instead of just one.

> My suggestion would be to add colibri support to "pan" (see #187, which is still incomplete and has been collecting dust, unfortunately) instead of building a separate high-level API for Colibri only.

Trying to unify the paths under a common API would only slow down the design of the colibri API. Furthermore, since I know nothing about PAN, integrating all of it into PAN would delay things even further, at least for me.